### PR TITLE
[RAPTOR-5781] update generic_predictor.py

### PR DESCRIPTION
**Pull request by timsetsfire**
Tied to bug report https://datarobot.atlassian.net/browse/RAPTOR-5781

Transform hook can return either a tuple or a dataframe -> updated generic_predictor.py accordingly.  

# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py [hunk 1]
```diff
@@ -105,7 +106,13 @@ def _materialize(self, parent_data_objs, user_data):
             binary_data, mimetype = StructuredInputReadUtils.read_structured_input_file_as_binary(
                 input_filename
             )
-            transformed_df = self._predictor.transform(binary_data=binary_data, mimetype=mimetype)
+            transformed_output = self._predictor.transform(
```
Comment (rvorobii) [inline custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:109]: can we make `transform` always return either dataframe or tuple?
what is the reason it's sometimes dataframe and sometimes tuple

Comment (jmbannon) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:109]: its dependent on sending `y` to the transform. I agree though, we should just always make it a tuple

**Review by jmbannon — APPROVED** [review]

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py (hunk 1 above)
Comment (timsetsfire) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:109]: so i'm curious on how to move forward with this.  To make it always return a tuple would require changes to model_adapter.py and we could essentially undo everything that was done to generic_predictor.  

part of model_adapter to change 👇 
```
        if self._custom_hooks.get(CustomHooks.TRANSFORM):
            try:
                transform_hook = self._custom_hooks[CustomHooks.TRANSFORM]
                transform_params = signature(transform_hook).parameters
                if len(transform_params) == 3:
                    transform_out = transform_hook(data, model, target_data)
                elif len(transform_params) == 2:
                    transform_out = transform_hook(data, model)
                else:
                    raise ValueError(
                        "Transform hook must take 2 or 3 arguments; "
                        "hook provided takes {}".format(len(transform_params))
                    )
                if type(transform_out) == tuple:
                    output_data, output_target = transform_out
                else:
                    output_data = transform_out
                    output_target = target_data

```
@yakov-g thoughts??

Comment (yakov-g) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:109]: So, as I understand, this solution won't work as `transform` which is defined here,
https://github.com/datarobot/datarobot-user-models/blob/a91996cc2674162424924cffe6e4de97e75a03e6/custom_model_runner/datarobot_drum/drum/language_predictors/base_language_predictor.py#L123-L128
will always return tuple, returned all the way from model_adapter (even if the hook returns a single value). So we already always return tuple. 
Thus, looks like, all that has to be done is adding `[0]`: 
`transformed_df = self._predictor.transform(binary_data=binary_data, mimetype=mimetype)[0]`

**Review by rvorobii — COMMENTED** [review]
I'm confused -- still don't understand if always returning a tuple is doable (which to me seems as a proper solution).
@jmbannon @yakov-g you're more familiar with this code, leaving this to you guys to find a proper solution.

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py [hunk 2]
```diff
@@ -105,7 +106,13 @@ def _materialize(self, parent_data_objs, user_data):
...
+                binary_data=binary_data, mimetype=mimetype
+            )
+            if type(transformed_output) == tuple:
+                transformed_df = transformed_output[0]
+            elif type(transformed_output) == DataFrame:
+                transfromed_df = transformed_output
```
Comment (meltzg) [inline custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:115]:
```suggestion
            if isinstance(transformed_output, tuple):
                transformed_df = transformed_output[0]
            elif isinstance(transformed_output, DataFrame):
                transfromed_df = transformed_output
```

**Review by meltzg — CHANGES_REQUESTED** [review]
could you add a test?

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py (hunk 2 above)
Comment (timsetsfire) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py]: changed.  

need to figure out what to test here.  @yakov-g any thoughts?

Comment (yakov-g) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py]: The test could be to check cases when transformed_output is either tuple or DataFrame, but, as I said before, looks like PythonPredictor always returns tuple. And transform is not implemented for other languages. 

So I cant think of any new test cases.

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py [hunk 3]
```diff
@@ -105,7 +106,13 @@ def _materialize(self, parent_data_objs, user_data):
             binary_data, mimetype = StructuredInputReadUtils.read_structured_input_file_as_binary(
                 input_filename
             )
-            transformed_df = self._predictor.transform(binary_data=binary_data, mimetype=mimetype)
+            transformed_output = self._predictor.transform(
+                binary_data=binary_data, mimetype=mimetype
+            )
+            if isinstance(transformed_output, tuple):
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py:112]: So just once again: do you agree that `transformed_output` is always a tuple? Why then we check for DataFrame type?

Comment (timsetsfire) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py]: @yakov-g finally address this and agree output is always a tuple and eliminated hanlding of dataframe.

**Review by yakov-g — APPROVED** [review]

File: custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py (hunk 2 above)
Comment (timsetsfire) [reply custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py]: resolve -> remove checks.  I agree that the PythonPredictor always returns a tuple.

**Review by meltzg — APPROVED** [review]

**Pull request merged**